src/packages/database/postgres/site-license/hook.ts

```diff
@@ -34,218 +39,309 @@ interface License {
 
 type LicenseMap = TypedMap<License>;
 
-async function get_valid_licenses(db): Promise<Map<string, LicenseMap>> {
-  // Todo -- filter on expiration...
-  if (licenses == null) {
-    licenses = await callback2(db.synctable.bind(db), {
-      table: "site_licenses",
-      columns: [
-        "title",
-        "expires",
-        "activates",
-        "upgrades",
-        "quota",
-        "run_limit",
-      ],
-      // TODO: Not bothing with the where condition will be fine up to a few thousand (?) site
-      // licenses, but after that it could take nontrivial time/memory during hub startup.
-      // So... this is a ticking time bomb.
-      //, where: { expires: { ">=": new Date() }, activates: { "<=": new Date() } }
-    });
-  }
-  return licenses.get();
-}
-
-/*
-Call this any time about to *start* the project.
-
-Check for site licenses, then set the site_license field for this project.
-The *value* for each key records what the license provides and whether or
-not it is actually being used by the project.
-
-If the license provides nothing new compared to what is already provided
-by already applied **licenses** and upgrades, then the license is *not*
-applied.   See
-      https://github.com/sagemathinc/cocalc/issues/4979
-*/
+// used to throttle lase_used updates per license
+const LAST_USED: { [license_id: string]: number } = {};
 
+/**
+ * Call this any time about to *start* the project.
+ *
+ * Check for site licenses, then set the site_license field for this project.
+ * The *value* for each key records what the license provides and whether or
+ * not it is actually being used by the project.
+ *
+ * If the license provides nothing new compared to what is already provided
+ * by already applied **licenses** and upgrades, then the license is *not*
+ * applied.
+ *
+ * related issues about it's heuristic:
+ * - https://github.com/sagemathinc/cocalc/issues/4979 -- do not apply a license if it does not provide upgrades
+ * - https://github.com/sagemathinc/cocalc/pull/5490 -- remove a license if it is expired
```

Comment:
FYI: I disagree with "remove a license if it is expired"...  We've had users and myself find it confusing.